### ISSUE TITLE
Questions about network architecture

### ISSUE THREAD

**Issue by lzj9072** [facebookresearch/wav2letter#401]
1. In the tutorial, the meaning of each dimension of TDS is defined as follow: 
![image](https://user-images.githubusercontent.com/26057369/64761336-70607800-d56e-11e9-90a7-48070c9294d6.png)
But in the architecture configuration, as the pic below, is it more reasonable that the meaning of each dim is TDS [channels][kernel width][input width][dropout prob]?
![image](https://user-images.githubusercontent.com/26057369/64761739-5f643680-d56f-11e9-9ab2-5ffc1be50934.png)
2. Is "LN: 3" means that the third dimension will be normalized?

**Comment by lzj9072** [comment]
@tlikhomanenko Thanks for replying. I have another question about LN...
In a mini-batch, I pad the training sample(audio) as the same size of the longest sample in that batch. Therefore the time sequence lengths of all batches may be different.
When using layer norm, there are two learnable parameters gamma and beta.
![image](https://user-images.githubusercontent.com/26057369/64944868-e6801a00-d8a1-11e9-95f1-58c2d19a684f.png)
As the seq2seq-tds paper says, _The layer normalization is over all dimensions for a given example including time_. Since the dimension of time is not fixed, the shapes of gamma and beta are not fixed. I wonder how to use LN when the length of time sequence is not fixed, just simply not using the learnable parameters?

**Comment by tlikhomanenko** [comment]
Hi @lzj9072,

in seq2seq-tds paper we do layer normalization across all dimensions except batch dim (so we do normalization across time too). Thus the sizes for gamma and beta will be only 1d vectors, which length is equal to the batch size, and then you do casting for other dimensions of the input of the LayerNorm.

For speech processing I would say if you are using some normalization of the layer output, you should do normalization at least across time because it is different for different batches (even for different samples).

**Comment by lzj9072** [comment]
Hi @tlikhomanenko,
I agree that normalization should be across time.

> Thus the sizes for gamma and beta will be only 1d vectors, which length is equal to the batch size

I cannot figure it out why the size of γ and β (parameters of layerNorm) are relevant with the batch size. Assume a tensor whose size is `[batch_size, channel, time, frequency]`, the LN should across the last three dimensions. Refer to the LN paper and the LN implement of pytorch, the size of γ and β should be` [channel, time, frequency]`.
I wonder how the layer norm can handle various lengths of audios when the input lengths determine the size of the LN parameters?

**Comment by tlikhomanenko** [comment]
Hi @lzj9072,

Could send reference? As I know if you do LN across last three dimensions, this means that you compute average across these dimensions or you reduce these dimensions (or do sum over indices i,j,k where i=1..D2 j=1..D3 k=1..D4) and the output will be vector [batch, 1, 1, 1]. In case of flashlight implementation you have gamma and beta with sizes corresponding to the sizes of the output, so their sizes will be [batch, 1, 1, 1]. Parameters of gamma and beta make sense of new mean and var that is why their sizes should be the same as sizes of computed mean and var, so [batch, 1, 1, 1]. I believe in pytorch and for BatchNorm and LayerNorm paper this will be the same as I describe.
